File: Desktop/GB/backend/models/cattle.py
from database import db
from datetime import datetime

class Cattle(db.Model):
    __tablename__ = 'cattle'
    
    id = db.Column(db.Integer, primary_key=True)
    tag_number = db.Column(db.String(50), unique=True, nullable=False)
    name = db.Column(db.String(100), nullable=False)
    breed = db.Column(db.String(50), nullable=False)
    date_of_birth = db.Column(db.Date, nullable=False)
    gender = db.Column(db.String(10), nullable=False)  # Male/Female
    weight = db.Column(db.Float, nullable=True)
    health_status = db.Column(db.String(50), default='Healthy')
    location = db.Column(db.String(100), nullable=True)
    purchase_date = db.Column(db.Date, nullable=True)
    purchase_price = db.Column(db.Float, nullable=True)
    current_status = db.Column(db.String(20), default='Active')  # Active/Sold/Deceased
    notes = db.Column(db.Text, nullable=True)
    created_at = db.Column(db.DateTime, default=datetime.utcnow)
    updated_at = db.Column(db.DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    
    # Relationships
    milk_records = db.relationship('MilkProduction', backref='cattle', lazy=True, cascade='all, delete-orphan')
    feeding_records = db.relationship('Feeding', backref='cattle', lazy=True, cascade='all, delete-orphan')
    
    def __repr__(self):
        return f'<Cattle {self.tag_number}: {self.name}>'
    
    def to_dict(self):
        return {
            'id': self.id,
            'tag_number': self.tag_number,
            'name': self.name,
            'breed': self.breed,
            'date_of_birth': self.date_of_birth.isoformat() if self.date_of_birth else None,
            'gender': self.gender,
            'weight': self.weight,
            'health_status': self.health_status,
            'location': self.location,
            'purchase_date': self.purchase_date.isoformat() if self.purchase_date else None,
            'purchase_price': self.purchase_price,
            'current_status': self.current_status,
            'notes': self.notes,
            'created_at': self.created_at.isoformat(),
            'updated_at': self.updated_at.isoformat()
        }
    
    def calculate_age_in_months(self):
        if self.date_of_birth:
            today = datetime.now().date()
            return (today.year - self.date_of_birth.year) * 12 + (today.month - self.date_of_birth.month)
        return None

File: Desktop/GB/cattle-management-mobile/backend-mongo/models/Cattle.js
const mongoose = require('mongoose');

const cattleSchema = new mongoose.Schema({
  tag_number: {
    type: String,
    required: true,
    unique: true,
    trim: true
  },
  name: {
    type: String,
    required: true,
    trim: true
  },
  breed: {
    type: String,
    required: true,
    enum: ['Holstein', 'Jersey', 'Angus', 'Hereford', 'Brahman', 'Simmental', 'Charolais', 'Other']
  },
  date_of_birth: {
    type: Date,
    required: true
  },
  gender: {
    type: String,
    required: true,
    enum: ['Male', 'Female']
  },
  weight: {
    type: Number,
    min: 0
  },
  health_status: {
    type: String,
    default: 'Healthy',
    enum: ['Healthy', 'Sick', 'Injured', 'Pregnant', 'Recovering']
  },
  location: {
    type: String,
    trim: true
  },
  purchase_date: {
    type: Date
  },
  purchase_price: {
    type: Number,
    min: 0
  },
  current_status: {
    type: String,
    default: 'Active',
    enum: ['Active', 'Sold', 'Deceased', 'Quarantined']
  },
  notes: {
    type: String,
    trim: true
  }
}, {
  timestamps: { createdAt: 'created_at', updatedAt: 'updated_at' }
});

// Indexes for better query performance
cattleSchema.index({ tag_number: 1 });
cattleSchema.index({ current_status: 1 });
cattleSchema.index({ health_status: 1 });
cattleSchema.index({ breed: 1 });

// Virtual for age calculation
cattleSchema.virtual('age_in_months').get(function() {
  if (!this.date_of_birth) return null;
  const today = new Date();
  const birth = this.date_of_birth;
  return (today.getFullYear() - birth.getFullYear()) * 12 + (today.getMonth() - birth.getMonth());
});

// Ensure virtual fields are serialized
cattleSchema.set('toJSON', { virtuals: true });

module.exports = mongoose.model('Cattle', cattleSchema);
